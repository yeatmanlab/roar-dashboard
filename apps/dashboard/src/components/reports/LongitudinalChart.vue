<template>
  <div class="longitudinal-chart">
    <PvChart
      type="line"
      :data="chartData"
      :options="chartOptions"
    />
  </div>
</template>

<script setup>
import { computed } from 'vue';
import PvChart from 'primevue/chart';
import _startCase from 'lodash/startCase';

const props = defineProps({
  longitudinalData: {
    type: Array,
    required: true
  },
  taskId: {
    type: String,
    required: true
  },
  grade: {
    type: Number,
    required: true
  }
});

const formatDate = (date) => {
  return new Date(date).toLocaleDateString(undefined, { 
    year: 'numeric',
    month: 'short',
    day: 'numeric'
  });
};

const formatScoreType = (type) => {
  return _startCase(type.replace(/([A-Z])/g, ' $1').toLowerCase());
};

// Helper function to get consistent colors for score types
const getColorForType = (type) => {
  const colorMap = {
    rawScore: '#2196F3',       // Blue
    percentile: '#4CAF50',     // Green
    standardScore: '#FF9800',  // Orange
    default: '#9C27B0'         // Purple
  };
  return colorMap[type] || colorMap.default;
};

// Prepare chart data
const chartData = computed(() => {
  console.log('LongitudinalChart - Input data:', props.longitudinalData);
  
  if (!props.longitudinalData?.length) {
    console.log('No longitudinal data available');
    return { labels: [], datasets: [] };
  }

  const sortedData = [...props.longitudinalData].sort((a, b) => new Date(a.date) - new Date(b.date));
  console.log('Sorted data:', sortedData);
  
  const labels = sortedData.map(entry => formatDate(entry.date));
  console.log('Labels:', labels);
  
  // Create datasets for each score type
  const datasets = [];
  
  // Define the score types we want to show
  const scoreTypes = ['rawScore', 'percentile', 'standardScore'];

  // Create a dataset for each score type
  scoreTypes.forEach(scoreType => {
    console.log(`Processing scores for ${scoreType}:`);
    const scores = sortedData.map(entry => {
      const score = entry.scores?.[scoreType];
      console.log(`- Entry scores:`, entry.scores, `Score for ${scoreType}:`, score);
      return score || null;
    });

    console.log(`Final scores for ${scoreType}:`, scores);

    // Only add the dataset if we have at least one valid score
    if (scores.some(score => score !== null)) {
      datasets.push({
        label: formatScoreType(scoreType),
        data: scores,
        fill: false,
        tension: 0.4,
        borderColor: getColorForType(scoreType),
        pointRadius: 4,
        pointHoverRadius: 6
      });
    }
  });

  console.log('Final datasets:', datasets);
  return { labels, datasets };
});

const chartOptions = computed(() => ({
  responsive: true,
  maintainAspectRatio: false,
  plugins: {
    legend: {
      position: 'bottom',
      labels: {
        usePointStyle: true,
        padding: 15
      }
    },
    tooltip: {
      mode: 'index',
      intersect: false,
      callbacks: {
        label: (context) => {
          const label = context.dataset.label || '';
          const value = context.parsed.y;
          return `${label}: ${Math.round(value)}`;
        }
      }
    }
  },
  scales: {
    y: {
      beginAtZero: true,
      grid: {
        color: 'rgba(0,0,0,0.1)'
      }
    },
    x: {
      grid: {
        display: false
      }
    }
  },
  interaction: {
    mode: 'nearest',
    axis: 'x',
    intersect: false
  }
}));
</script>

<style scoped>
.longitudinal-chart {
  height: 300px;
  margin: 1rem 0;
  padding: 1rem;
  background-color: white;
  border-radius: 8px;
  box-shadow: 0 1px 3px rgba(0,0,0,0.1);
}
</style>
