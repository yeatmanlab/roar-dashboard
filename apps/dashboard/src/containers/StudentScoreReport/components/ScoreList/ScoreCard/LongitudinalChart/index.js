export { default as LongitudinalChartScreen } from './LongitudinalChart.vue';
export { default as LongitudinalChartPrint } from './LongitudinalChart.print.vue';
