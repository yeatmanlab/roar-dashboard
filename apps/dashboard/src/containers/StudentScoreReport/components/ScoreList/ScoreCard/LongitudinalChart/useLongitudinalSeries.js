import { computed } from 'vue';
import { getDialColor, supportLevelColors } from '@/helpers/reports';

const SCORE_TYPES = {
  rawScore: { key: 'rawScore', label: 'Raw Score', priority: 1 },
  percentile: { key: 'percentile', label: 'Percentile', priority: 2 },
  standardScore: { key: 'standardScore', label: 'Standard Score', priority: 3 },
};

const getLabelByScoreType = (t) => SCORE_TYPES[t]?.label ?? 'Score';
const preferredTypes = Object.values(SCORE_TYPES)
  .sort((a, b) => a.priority - b.priority)
  .map((t) => t.key);

export function useLongitudinalSeries(props) {
  const sorted = computed(() => {
    const a = props.longitudinalData || [];
    return [...a].sort((x, y) => new Date(x.date) - new Date(y.date));
  });

  const chosenType = computed(() => preferredTypes.find((t) => sorted.value.some((e) => e.scores?.[t] != null)));

  const series = computed(() => {
    if (!chosenType.value) return [];
    const t = chosenType.value;
    return sorted.value
      .filter((e) => e.scores?.[t] != null && !Number.isNaN(+e.scores[t]))
      .map((e) => {
        const x = new Date(e.date);
        const y = +e.scores[t];

        const rawScore = e.scores.rawScore;
        const percentile = e.scores.percentileScore ?? e.scores.percentile ?? null;
        const standardScore = e.scores.standardScore ?? null;

        const color = getDialColor(props.studentGrade, percentile, rawScore, props.taskId);
        console.log('[DEBUG] Point:', {
          taskId: props.taskId,
          grade: props.studentGrade,
          percentile,
          rawScore,
          date: x.toISOString(),
          score: y,
          colorFromDialColor: color,
          finalColor: color || supportLevelColors.Assessed,
        });
        return {
          x,
          y,
          assignmentId: e.assignmentId || e.administrationId || '',
          percentile,
          standardScore,
          color: color || supportLevelColors.Assessed,
        };
      });
  });

  const seriesLabel = computed(() => getLabelByScoreType(chosenType.value));
  const seriesStroke = computed(() => supportLevelColors.Assessed);

  const xDomain = computed(() => {
    if (!series.value.length) return [new Date(), new Date()];
    return [series.value[0].x, series.value[series.value.length - 1].x];
  });

  const yDomain = computed(() => {
    if (!series.value.length) return [0, 1];
    const ys = series.value.map((p) => p.y);
    const min = Math.min(...ys, 0);
    const max = Math.max(...ys);
    return [min, max === min ? min + 1 : max];
  });

  return { series, seriesLabel, seriesStroke, xDomain, yDomain };
}
