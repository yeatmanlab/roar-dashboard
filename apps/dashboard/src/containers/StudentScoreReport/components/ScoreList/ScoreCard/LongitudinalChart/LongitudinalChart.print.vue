<template>
  <div class="px-2 pt-2 rounded border border-gray-100 border-solid">
    <canvas v-if="showCanvas" ref="canvasRef" class="w-full h-full"></canvas>
    <img v-else :src="imgSrc" class="w-full h-auto" alt="Longitudinal chart" />
  </div>
</template>

<script setup>
import { computed, ref, onMounted, onBeforeUnmount, nextTick } from 'vue';
import Chart from 'chart.js/auto';
import 'chartjs-adapter-date-fns';
import { useLongitudinalSeries } from './useLongitudinalSeries';

const props = defineProps({
  longitudinalData: { type: Array, required: true },
  taskId: { type: String, required: true },
  studentGrade: { type: String, required: true },
});

const { series, seriesLabel } = useLongitudinalSeries(props);

const canvasRef = ref(null);
let chartInstance = null;
const showCanvas = ref(true);
const imgSrc = ref('');

const chartData = computed(() => ({
  datasets: [
    {
      label: seriesLabel.value,
      data: series.value.map((p) => ({ x: p.x, y: p.y })),
      borderColor: 'rgba(0,0,0,0.85)',
      backgroundColor: 'rgba(0,0,0,0.85)',
      pointBackgroundColor: series.value.map((p) => p.color),
      pointBorderColor: series.value.map((p) => p.color),
      borderWidth: 2,
      pointRadius: 5,
      pointHoverRadius: 5,
      tension: 0.4,
      spanGaps: true,
    },
  ],
}));

const chartOptions = computed(() => ({
  responsive: true,
  maintainAspectRatio: false,
  devicePixelRatio: 3,
  animation: false,
  events: [],
  plugins: { legend: false, tooltip: false },
  scales: {
    y: {
      beginAtZero: true,
      ticks: { color: '#000' },
      grid: { color: 'rgba(0,0,0,0.15)' },
    },
    x: {
      type: 'timeseries',
      time: { unit: 'month', displayFormats: { month: 'MMM yyyy' } },
      ticks: { color: '#000', maxRotation: 0, autoSkip: true, autoSkipPadding: 8 },
      grid: { display: false },
    },
  },
}));

async function renderThenSnapshot() {
  // Render the chart
  const ctx = canvasRef.value?.getContext('2d');
  if (!ctx) return;
  if (chartInstance) {
    chartInstance.destroy();
    chartInstance = null;
  }

  chartInstance = new Chart(ctx, {
    type: 'line',
    data: chartData.value,
    options: chartOptions.value,
  });

  // Ensure it has painted at least once before snapshotting
  await nextTick();
  await new Promise((r) => requestAnimationFrame(() => requestAnimationFrame(r)));

  // Snapshot to data URL and swap to <img>
  const url = chartInstance.toBase64Image('image/png', 1); // full quality
  imgSrc.value = url;
  showCanvas.value = false;

  // Cleanup
  chartInstance.destroy();
  chartInstance = null;
}

onMounted(renderThenSnapshot);
onBeforeUnmount(() => {
  chartInstance?.destroy();
  chartInstance = null;
});
</script>
