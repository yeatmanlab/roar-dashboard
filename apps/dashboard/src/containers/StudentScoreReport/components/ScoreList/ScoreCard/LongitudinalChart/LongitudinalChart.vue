<template>
  <div class="p-2 rounded border border-gray-100 border-solid" data-cy="longitudinal-chart">
    <div class="h-64">
      <canvas ref="canvasRef" class="w-full h-full"></canvas>
    </div>
    <div v-if="showSupportLevels" class="flex gap-3 justify-end mt-2 text-xs">
      <div class="flex gap-1 items-center">
        <div class="w-3 h-3 rounded-full" :style="{ backgroundColor: SCORE_SUPPORT_LEVEL_COLORS.ABOVE }"></div>
        <span>Achieved Skill</span>
      </div>
      <div class="flex gap-1 items-center">
        <div class="w-3 h-3 rounded-full" :style="{ backgroundColor: SCORE_SUPPORT_LEVEL_COLORS.SOME }"></div>
        <span>Developing Skill</span>
      </div>
      <div class="flex gap-1 items-center">
        <div class="w-3 h-3 rounded-full" :style="{ backgroundColor: SCORE_SUPPORT_LEVEL_COLORS.BELOW }"></div>
        <span>Needs Extra Support</span>
      </div>
    </div>
  </div>
</template>

<script setup>
import { computed, ref, onMounted, onBeforeUnmount, watch } from 'vue';
import Chart from 'chart.js/auto';
import 'chartjs-adapter-date-fns';
import { useLongitudinalSeries } from './useLongitudinalSeries';
import { SCORE_SUPPORT_LEVEL_COLORS } from '@/constants/scores';
import {
  tasksToDisplayPercentCorrect,
  tasksToDisplayTotalCorrect,
  tasksToDisplayGradeEstimate,
} from '@/helpers/reports';

const canvasRef = ref(null);
let chartInstance = null;

const props = defineProps({
  longitudinalData: { type: Array, required: true },
  taskId: { type: String, required: true },
  studentGrade: { type: String, required: true },
  currentAssignmentId: { type: String, required: true },
  scoreLabel: { type: String, required: true },
});

const { series, seriesStroke } = useLongitudinalSeries(props);

// Filter series to only show points up to current assignment
const filteredSeries = computed(() => {
  const currentAssignment = series.value.find((p) => p.assignmentId === props.currentAssignmentId);
  if (!currentAssignment) return [];
  const currentDate = currentAssignment.x;
  return series.value.filter((p) => p.x <= currentDate);
});

const showSupportLevels = computed(() => {
  // Don't show support levels if the task is in any of these lists
  const isDisplayTask = [
    ...tasksToDisplayPercentCorrect,
    ...tasksToDisplayTotalCorrect,
    ...tasksToDisplayGradeEstimate,
  ].includes(props.taskId);

  // Check if any data point has a support level color
  const hasSupportLevels = series.value.some((point) =>
    Object.values(SCORE_SUPPORT_LEVEL_COLORS).includes(point.color),
  );

  return !isDisplayTask && hasSupportLevels;
});

const chartData = computed(() => ({
  datasets: [
    {
      label: props.scoreLabel,
      data: filteredSeries.value.map((p) => ({ x: p.x, y: p.y })),
      tension: 0.4,
      borderColor: seriesStroke,
      pointRadius: filteredSeries.value.map((p) =>
        p.assignmentId && p.assignmentId === props.currentAssignmentId ? 8 : 4,
      ),
      pointHoverRadius: filteredSeries.value.map((p) =>
        p.assignmentId && p.assignmentId === props.currentAssignmentId ? 10 : 6,
      ),
      pointBackgroundColor: filteredSeries.value.map((p) => p.color),
      pointBorderColor: filteredSeries.value.map((p) =>
        p.assignmentId && p.assignmentId === props.currentAssignmentId ? '#000000' : p.color,
      ),
      pointBorderWidth: filteredSeries.value.map((p) =>
        p.assignmentId && p.assignmentId === props.currentAssignmentId ? 2 : 1,
      ),
      pointStyle: filteredSeries.value.map((p) =>
        p.assignmentId && p.assignmentId === props.currentAssignmentId ? 'rectRot' : 'circle',
      ),
      spanGaps: true,
    },
  ],
}));

const chartOptions = computed(() => ({
  responsive: true,
  normalized: true,
  maintainAspectRatio: false,
  plugins: {
    legend: false,
    tooltip: {
      mode: 'index',
      intersect: false,
      callbacks: {
        title: (items) => {
          const ts = items[0].parsed.x;
          return new Intl.DateTimeFormat(undefined, { year: 'numeric', month: 'short', day: 'numeric' }).format(ts);
        },
        label: (ctx) => {
          const p = series.value[ctx.dataIndex];
          const lines = [`${props.scoreLabel}: ${p.y}`];
          if (p.percentile != null) lines.push(`Percentile: ${p.percentile}`);
          if (p.standardScore != null) lines.push(`Standard Score: ${p.standardScore}`);
          if (p.assignmentId === props.currentAssignmentId) lines.unshift('✦ Current Score Report ✦');
          return lines;
        },
      },
    },
  },
  scales: {
    y: {
      beginAtZero: true,
      grid: { color: 'rgba(0,0,0,0.1)' },
      title: {
        display: true,
        text: props.scoreLabel,
        font: {
          size: 12,
        },
      },
    },
    x: {
      type: 'time',
      time: { unit: 'month', displayFormats: { month: 'MMM yyyy' } },
      ticks: { maxRotation: 0, autoSkip: true, autoSkipPadding: 8 },
      grid: { display: false },
    },
  },
  interaction: { mode: 'nearest', axis: 'x', intersect: false },
}));

function createChart() {
  const ctx = canvasRef.value?.getContext('2d');
  if (!ctx) return;
  if (chartInstance) chartInstance.destroy();
  chartInstance = new Chart(ctx, { type: 'line', data: chartData.value, options: chartOptions.value });
}

onMounted(createChart);
onBeforeUnmount(() => {
  chartInstance?.destroy();
  chartInstance = null;
});

watch([chartData, chartOptions], () => {
  if (!chartInstance) return;
  chartInstance.data = chartData.value;
  chartInstance.options = chartOptions.value;
  chartInstance.update();
});
</script>
