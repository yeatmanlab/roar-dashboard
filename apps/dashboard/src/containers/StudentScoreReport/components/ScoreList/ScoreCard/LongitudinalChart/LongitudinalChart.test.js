import { describe, it, expect, beforeEach, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import LongitudinalChart from './LongitudinalChart.vue';

// Mock roar-utils
vi.mock('@bdelab/roar-utils', async () => {
  const actual = await vi.importActual('@bdelab/roar-utils');
  return {
    ...actual,
    getGrade: vi.fn(),
  };
});

// Mock helpers/reports
vi.mock('@/helpers/reports', () => ({
  getDialColor: vi.fn(),
  supportLevelColors: {
    above: 'green',
    some: '#edc037',
    below: '#c93d82',
  },
}));

// Mock Chart.js
vi.mock('chart.js/auto', () => ({
  default: vi.fn().mockImplementation(() => ({
    destroy: vi.fn(),
    update: vi.fn(),
    data: {},
    options: {},
  })),
}));

// Mock the composable
vi.mock('./useLongitudinalSeries', () => ({
  useLongitudinalSeries: vi.fn(() => ({
    series: {
      value: [
        {
          x: new Date('2024-01-01'),
          y: 45,
          color: '#4CAF50',
          percentile: 65,
          standardScore: 95,
        },
        {
          x: new Date('2024-06-01'),
          y: 50,
          color: '#4CAF50',
          percentile: 75,
          standardScore: 105,
        },
      ],
    },
    seriesLabel: {
      value: 'Raw Score',
    },
    seriesStroke: {
      value: '#2196F3',
    },
  })),
}));

// Mock HTMLCanvasElement.prototype.getContext
HTMLCanvasElement.prototype.getContext = vi.fn(() => ({
  fillStyle: '',
  fillRect: vi.fn(),
  clearRect: vi.fn(),
  getImageData: vi.fn(),
  putImageData: vi.fn(),
  createImageData: vi.fn(),
  setTransform: vi.fn(),
  drawImage: vi.fn(),
  save: vi.fn(),
  restore: vi.fn(),
  beginPath: vi.fn(),
  moveTo: vi.fn(),
  lineTo: vi.fn(),
  closePath: vi.fn(),
  stroke: vi.fn(),
  translate: vi.fn(),
  scale: vi.fn(),
  rotate: vi.fn(),
  arc: vi.fn(),
  fill: vi.fn(),
  measureText: vi.fn(() => ({ width: 0 })),
  transform: vi.fn(),
  rect: vi.fn(),
  clip: vi.fn(),
}));

describe('LongitudinalChart.vue', () => {
  const defaultProps = {
    longitudinalData: [
      { date: '2024-01-01', scores: { rawScore: 45, percentile: 65 } },
      { date: '2024-06-01', scores: { rawScore: 50, percentile: 75 } },
    ],
    taskId: 'swr',
    studentGrade: '3',
    currentAssignmentId: 'test-id',
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe('snapshots', () => {
    it('renders with longitudinal data', () => {
      const wrapper = mount(LongitudinalChart, {
        props: defaultProps,
      });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders with single data point', () => {
      const wrapper = mount(LongitudinalChart, {
        props: {
          ...defaultProps,
          longitudinalData: [{ date: '2024-01-01', scores: { rawScore: 45 } }],
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders with different task', () => {
      const wrapper = mount(LongitudinalChart, {
        props: {
          ...defaultProps,
          taskId: 'sre',
        },
      });

      expect(wrapper.html()).toMatchSnapshot();
    });
  });

  describe('component behavior', () => {
    it('renders canvas element', () => {
      const wrapper = mount(LongitudinalChart, {
        props: defaultProps,
      });

      const canvas = wrapper.find('canvas');
      expect(canvas.exists()).toBe(true);
    });

    it('has correct container styling', () => {
      const wrapper = mount(LongitudinalChart, {
        props: defaultProps,
      });

      const container = wrapper.find('.rounded');
      expect(container.exists()).toBe(true);
      expect(container.classes()).toContain('border');
      expect(container.classes()).toContain('border-gray-100');
    });

    it('mounts with required props', () => {
      const wrapper = mount(LongitudinalChart, {
        props: defaultProps,
      });

      expect(wrapper.props('longitudinalData')).toEqual(defaultProps.longitudinalData);
      expect(wrapper.props('taskId')).toBe('swr');
      expect(wrapper.props('studentGrade')).toBe('3');
    });

    it('canvas has proper classes', () => {
      const wrapper = mount(LongitudinalChart, {
        props: defaultProps,
      });

      const canvas = wrapper.find('canvas');
      expect(canvas.classes()).toContain('w-full');
      expect(canvas.classes()).toContain('h-full');
    });
  });

  describe('chart lifecycle', () => {
    it('creates chart on mount', () => {
      const wrapper = mount(LongitudinalChart, {
        props: defaultProps,
      });

      // Chart should be created (mocked)
      expect(wrapper.vm).toBeDefined();
    });

    it('destroys chart on unmount', () => {
      const wrapper = mount(LongitudinalChart, {
        props: defaultProps,
      });

      wrapper.unmount();

      // Verify component was unmounted
      expect(wrapper.exists()).toBe(false);
    });
  });
});
