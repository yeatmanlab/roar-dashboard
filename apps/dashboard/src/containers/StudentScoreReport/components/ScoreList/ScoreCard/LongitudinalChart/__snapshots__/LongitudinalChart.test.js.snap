// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LongitudinalChart.vue > snapshots > renders with different task 1`] = `
"<div class="p-2 rounded border border-gray-100 border-solid" data-cy="longitudinal-chart">
  <div class="h-64"><canvas class="w-full h-full"></canvas></div>
  <!--v-if-->
</div>"
`;

exports[`LongitudinalChart.vue > snapshots > renders with longitudinal data 1`] = `
"<div class="p-2 rounded border border-gray-100 border-solid" data-cy="longitudinal-chart">
  <div class="h-64"><canvas class="w-full h-full"></canvas></div>
  <!--v-if-->
</div>"
`;

exports[`LongitudinalChart.vue > snapshots > renders with single data point 1`] = `
"<div class="p-2 rounded border border-gray-100 border-solid" data-cy="longitudinal-chart">
  <div class="h-64"><canvas class="w-full h-full"></canvas></div>
  <!--v-if-->
</div>"
`;
